google-cloud-pubsub/test/google/cloud/pubsub/flow_controller_test.rb

```diff
@@ -0,0 +1,545 @@
+    end
+
+    it "raises when a single message is > message_limit" do
+      expect do
+        flow_controller = Google::Cloud::PubSub::FlowController.new(
+          message_limit: 0, # Non-sane setting
+          byte_limit: 10_000_000,
+          limit_exceeded_behavior: :block
+        )
+      end.must_raise ArgumentError
+    end
+
+    it "raises when a single message is > byte_limit" do
+      flow_controller = Google::Cloud::PubSub::FlowController.new(
+        message_limit: 1000,
+        byte_limit: 3,
+        limit_exceeded_behavior: :block
+      )
+
+      expect do
+        flow_controller.acquire 3 * 2
+      end.must_raise Google::Cloud::PubSub::FlowControlLimitError
+    end
+
+    it "blocks when > message_limit" do
+      flow_controller = Google::Cloud::PubSub::FlowController.new(
+        message_limit: 1,
+        byte_limit: 100,
+        limit_exceeded_behavior: :block
+      )
+
+      adding_1_done = Concurrent::Event.new
+      adding_2_done = Concurrent::Event.new
+      adding_3_done = Concurrent::Event.new
+      releasing_1_done = Concurrent::Event.new
+      releasing_2_done = Concurrent::Event.new
+      releasing_3_done = Concurrent::Event.new
+
+      _(flow_controller.outstanding_bytes).must_equal 0
+
+      run_in_thread flow_controller, :acquire, 3, adding_1_done
+      assert adding_1_done.wait(0.1), "Adding message 1 never unblocked."
+      _(flow_controller.outstanding_bytes).must_equal 3 # Implementation detail
+
+      run_in_thread flow_controller, :acquire, 3, adding_2_done
+      refute adding_2_done.wait(0.1), "Adding message 2 did not block."
+      _(flow_controller.outstanding_bytes).must_equal 3 # Implementation detail
+
+      run_in_thread flow_controller, :acquire, 3, adding_3_done
+      refute adding_3_done.wait(0.1), "Adding message 3 did not block."
+      _(flow_controller.outstanding_bytes).must_equal 3 # Implementation detail
+
+      run_in_thread flow_controller, :release, 3, releasing_1_done
+      assert releasing_1_done.wait(0.1), "Releasing message 1 errored."
+
+      assert adding_2_done.wait(0.1), "Adding message 2 never unblocked."
+      _(flow_controller.outstanding_bytes).must_equal 3 # Implementation detail
+
+      run_in_thread flow_controller, :release, 3, releasing_2_done
+      assert releasing_2_done.wait(0.1), "Releasing message 2 errored."
+
+      assert adding_3_done.wait(0.1), "Adding message 3 never unblocked."
+      _(flow_controller.outstanding_bytes).must_equal 3 # Implementation detail
+
+      run_in_thread flow_controller, :release, 3, releasing_3_done
+      assert releasing_3_done.wait(0.1), "Releasing message 3 errored."
+      _(flow_controller.outstanding_bytes).must_equal 0 # Implementation detail
+    end
+
+    it "blocks when > byte_limit" do
+      flow_controller = Google::Cloud::PubSub::FlowController.new(
+        message_limit: 1000,
+        byte_limit: 3,
```

Comment:
Should we also have a test that checks the case where we don't exactly fill up the byte limit? (e.g. if this were 4, and the first acquisition leaves 1 left, the next acquisition of 3 would still block and not try any partial weirdness.)

Reply:
Added, thanks!